### PR TITLE
More Math Functions

### PR DESCRIPTION
**Pull request by TheAbsolutionism**
### Description
This PR aims to add more math functions into Skript such as:
- mean(numbers)
- median(numbers)
- factorial(number)
- root(number, number)
- permutation(number, number)
- combination(number, number)

Giving users the option to use these built in functions that handle all the math for them.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #6007

### CODE REVIEW

File: src/main/java/ch/njol/util/Math2.java [hunk 1]
```diff
@@ -163,4 +161,16 @@ public static long multiplyClamped(long x, long y) {
 		return result;
 	}
 
+	public static @Nullable Long factorial(long x) {
```
Comment (Moderocky) [inline src/main/java/ch/njol/util/Math2.java:164]: This is probably quite dangerous because it's very likely to overflow (for, like, most numbers) and there's no detection for that or warning to the user and their result will just be incorrect.

**Review by Moderocky — COMMENTED** [review]
You can do combination without factorial.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 2]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+				double total = 0;
+				for (Object object : params[0]) {
+					Number number = (Number) object;
+					total += number.doubleValue();
+				}
+				total /= params[0].length;
+				return new Number[]{total};
+			}
+		})
+			.description("Get the average of a list of numbers.")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:722]:
```suggestion
			.description("Get the mean (average) of a list of numbers.")
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 3]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+			@Override
+			public Number @Nullable [] executeSimple(Object[][] params) {
+				Number root = (Number) params[0][0];
+				if (root.intValue() == 0)
+					return null;
+				Number number = (Number) params[1][0];
+				return new Number[]{Math.pow(number.doubleValue(), (1 / root.doubleValue()))};
+			}
+		})
+			.description("Get the result of the 'number' being rooted to the power of 'root'")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:791]:
```suggestion
			.description("Calculates the <i>n</i>th root of a number.")
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 4]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+					return null;
+				return new Number[]{total};
+			}
+		})
+			.description()
+			.examples()
+			.since("INSERT VERSION");
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("root", new Parameter[]{
+			new Parameter<>("root", DefaultClasses.NUMBER, true, null),
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:779]:
```suggestion
			new Parameter<>("n", DefaultClasses.NUMBER, true, null),
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 5]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+			public Number @Nullable [] executeSimple(Object[][] params) {
+				Number total = (Number) params[0][0];
+				Number set = (Number) params[1][0];
+				long denominator = total.longValue() - set.longValue();
+				Long totalF = Math2.factorial(total.longValue());
+				Long denominatorF = Math2.factorial(denominator);
+				if (totalF == null || denominatorF == null)
+					return null;
+				Skript.adminBroadcast("Num: " + totalF);
+				Skript.adminBroadcast("Den: " + denominatorF);
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:813]:
```suggestion
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 6]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+				double result = totalF.doubleValue()/denominatorF.doubleValue();
+				return new Number[]{result};
+			}
+		})
+			.description(
+				"Get the number of possible ordered arrangements from numbers 1 to the provided 'total' within a 'set' size.",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:819]:
```suggestion
				"Get the number of possible ordered arrangements from 1 to 'options' with each arrangement having a size equal to 'size'",
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 7]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+			.examples(
+				"root(2, 4) = 2 # same as sqrt(4)",
+				"root(4, 16) = 2",
+				"root(-4, 16) = 0.5"
+			)
+			.since("INSERT VERSION");
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("permutation", new Parameter[]{
+			new Parameter<>("total", DefaultClasses.NUMBER, true, null),
+			new Parameter<>("set", DefaultClasses.NUMBER, true, null)
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:801]:
```suggestion
			new Parameter<>("options", DefaultClasses.NUMBER, true, null),
			new Parameter<>("size", DefaultClasses.NUMBER, true, null)
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 8]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+					return null;
+				Skript.adminBroadcast("Num: " + totalF);
+				Skript.adminBroadcast("Den: " + denominatorF);
+				double result = totalF.doubleValue()/denominatorF.doubleValue();
+				return new Number[]{result};
+			}
+		})
+			.description(
+				"Get the number of possible ordered arrangements from numbers 1 to the provided 'total' within a 'set' size.",
+				"For example,  A permutation of 3 with a set size of 1 returns 3: (1), (2), (3)",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:820]:
```suggestion
				"For example, a permutation with 3 options and an arrangement size of 1, returns 3: (1), (2), (3)",
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 9]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+				"A permutation of 3 with a set size of 2 returns 6: (1, 2), (1, 3), (2, 1), (2, 3), (3, 1), (3, 2)"
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:821]:
```suggestion
				"A permutation with 3 options and an arrangement size of 2 returns 6: (1, 2), (1, 3), (2, 1), (2, 3), (3, 1), (3, 2)"
```
they dont return sets as `{2, 3} = {3, 2}`, so this would only return 3

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 10]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+					Long setF = Math2.factorial(set.longValue());
+					Long denominatorF = Math2.factorial(denominator);
+					if (totalF == null || setF == null || denominatorF == null)
+						return null;
+					double result = totalF.doubleValue()/(denominatorF.doubleValue() * setF.doubleValue());
+					return new Number[]{result};
+				}
+			})
+			.description(
+				"Get the number of unique unordered arrangements from numbers 1 to the provided 'total' within a 'set' size.",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:849]:
```suggestion
				"Get the number of possible sets from 1 to 'options' with each set having a size equal to 'size'",
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 11]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+				"For example, A combination of 3 with a set size of 1 returns 3: (1), (2), (3)",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:850]:
```suggestion
				"For example, a combination with 3 options and a set size of 1, returns 3: (1), (2), (3)",
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 12]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+				"A combination of 3 with a set size of 2 returns 3: (1, 2), (1, 3), (2, 3)"
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:851]:
```suggestion
				"A combination of 3 options with a set size of 2 returns 3: (1, 2), (1, 3), (2, 3)"
```

File: src/test/skript/tests/syntaxes/functions/permutation.sk [hunk 13]
```diff
@@ -0,0 +1,3 @@
+test "permutation function":
+	assert permutation(10, 2) = 90 with "permutation of 10 with a set of 8 should be 90"
+	assert permutation(10, 4) = 5040 with "permutation of 10 with a set of 4 should be 5040"
```
Comment (Efnilite) [inline src/test/skript/tests/syntaxes/functions/permutation.sk:3]: big numbers

File: src/test/skript/tests/syntaxes/functions/factorial.sk [hunk 14]
```diff
@@ -0,0 +1,7 @@
+test "factorial function":
+	assert factorial(0) = 1 with "factorial of 0 should be 1"
+	assert factorial(1) = 1 with "factorial of 1 should be 1"
+	assert factorial(2) = 2 with "factorial of 2 should be 2"
+	assert factorial(3) = 6 with "factorial of 3 should be 6"
+	assert factorial(4) = 24 with "factorial of 4 should be 24"
+	assert factorial(5) = 120 with "factorial of 5 should be 120"
```
Comment (Efnilite) [inline src/test/skript/tests/syntaxes/functions/factorial.sk:7]: im pretty sure it was discussed that this is a bad idea

File: src/test/skript/tests/syntaxes/functions/combination.sk [hunk 15]
```diff
@@ -0,0 +1,3 @@
+test "combination function":
+	assert combination(10, 8) = 45 with "combination of 10 with a set of 8 should be 45"
+	assert combination(5, 3) = 10 with "combination of 5 with a set of 3 should be 10"
```
Comment (Efnilite) [inline src/test/skript/tests/syntaxes/functions/combination.sk:3]: big numbers

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 9 above)
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:821]: Thats combination

File: src/test/skript/tests/syntaxes/functions/factorial.sk (hunk 14 above)
Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/functions/factorial.sk:7]: Oh? What's wrong with it?

Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/functions/factorial.sk:7]: Not so much that it was a bad idea, but that it wasn't very useful to add. You can only go up to about 20! before you lose precision by switching to doubles, which lets you get to 170! but only as approximations.

Adding on to this, the factorial function itself is quite niche in terms of use cases, mainly for combinatorics which can be calculated without factorials. All that together means it was concluded that if a user wants factorial, they can just write that function itself since it's so simple to code, anyway.

Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/functions/factorial.sk:7]: Ah ok, so just remove factorial then? Atleast the `registerFunction` since permutation and combination require the `Math2.factorial` I added

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 16]
```diff
@@ -711,6 +704,159 @@ public String[] executeSimple(Object[][] params) {
...
+				if (total == null)
+					return null;
+				return new Number[]{total};
+			}
+		})
+			.description()
+			.examples()
+			.since("INSERT VERSION");
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("root", new Parameter[]{
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:778]: What's the point of this one? We already have a exponent operator:
```
x ^ (1/2)
```

File: src/test/skript/tests/syntaxes/functions/combination.sk [hunk 17]
```diff
@@ -0,0 +1,3 @@
+test "combination function":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/functions/combination.sk:1]: all these tests need edge case tests, so with nan, infinity, null values, really large numbers, etc.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/functions/factorial.sk (hunk 14 above)
Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/functions/factorial.sk:7]: Those can and should be calculated without factorials, since your factorial function only supports 20! at maximum.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 16 above)
Comment (Moderocky) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:778]: A lot of people probably don't know this.

File: src/test/skript/tests/syntaxes/functions/factorial.sk (hunk 14 above)
Comment (Moderocky) [reply src/test/skript/tests/syntaxes/functions/factorial.sk:7]: I don't mind there being a factorial function as long as it has good documentation about values it won't work for and prevents overflow, I just don't want something that behaves in a way users don't expect and can't really account for.

Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/functions/factorial.sk:7]: I mean, I can just have it be a Double instead of a Long, that way it returns Infinity and/or NaN

Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/functions/factorial.sk:7]: well generally it's not all that useful to have approximate factorials, since that will throw off the combination/permutation calculations if you use it in that case. I don't mind the factorial function existing, but it shouldn't be relied on if it's using floating point numbers.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 16 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:778]: I guess that's fair, yeah

**Comment by TheAbsolutionism** [comment]
Errors are not cause of me FYI.

**Comment by sovdeeth** [comment]
> Errors are not cause of me FYI.

yes, they are. `Infinity` is not valid, you should be using `infinity value`

**Comment by TheAbsolutionism** [comment]
> yes, they are. `Infinity` is not valid, you should be using `infinity value`

Darn, I thought I was using it right and that the one issue was messing it up

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 18]
```diff
@@ -711,6 +704,209 @@ public String[] executeSimple(Object[][] params) {
...
+					return new Number[]{Math.sqrt(number)};
+				}
+				return new Number[]{Math.pow(number, (1 / n))};
+			}
+		})
+			.description("Calculates the <i>n</i>th root of a number.")
+			.examples(
+				"root(2, 4) = 2 # same as sqrt(4)",
+				"root(4, 16) = 2",
+				"root(-4, 16) = 0.5"
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:828]:
```suggestion
				"root(-4, 16) = 0.5 # same as 16^(-1/4)"
```

File: src/test/skript/tests/syntaxes/functions/median.sk [hunk 19]
```diff
@@ -0,0 +1,6 @@
+test "median function":
```
Comment (Efnilite) [inline src/test/skript/tests/syntaxes/functions/median.sk:1]: invalid nums here as well maybe

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 20]
```diff
@@ -711,6 +705,226 @@ public String[] executeSimple(Object[][] params) {
...
+			new Parameter<>("numbers", DefaultClasses.NUMBER, false, null)
+		}, DefaultClasses.NUMBER, true) {
+			@Override
+			public Number @Nullable [] executeSimple(Object[][] params) {
+				double total = 0;
+				for (Object object : params[0]) {
+					Number number = (Number) object;
+					if (number instanceof Double doubleNumber && (doubleNumber.isInfinite() || doubleNumber.isNaN())) {
+						Skript.error("You cannot get the mean of a set of number that includes 'infinity' or 'nan'.");
+						return null;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:719]: You can't skript.error here, this is runtime, not parsetime.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 21]
```diff
@@ -711,6 +705,226 @@ public String[] executeSimple(Object[][] params) {
...
+					})
+					.map(object -> (Number) object)
+					.sorted(((o1, o2) -> {
+						Double n1 = o1.doubleValue();
+						Double n2 = o2.doubleValue();
+						return n1.compareTo(n2);
+					}))
+					.toArray(Number[]::new);
+				if (invalid.get()) {
+					Skript.error("You cannot get the median of a set of numbers that includes 'infinity' or 'nan'.");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:759]: again, the above applies to all

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 20 above)
Comment (Efnilite) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:719]: maybe it should just skip infinity and nans

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 22]
```diff
@@ -711,6 +705,226 @@ public String[] executeSimple(Object[][] params) {
...
+						break;
+					result *= i;
+				}
+				if (negative)
+					result *= -1d;
+				return new Number[]{result};
+			}
+		})
+			.description(
+				"Get the factorial of a number. "
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:812]: maybe add that this shouldnt be used to calculate combinations or permutations

**Review by Efnilite — APPROVED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 23]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+			@Override
+			public Number @Nullable [] executeSimple(Object[][] params) {
+				double total = 0;
+				for (Object object : params[0]) {
+					Number number = (Number) object;
+					if (number instanceof Double doubleNumber && (doubleNumber.isInfinite() || doubleNumber.isNaN()))
+						return null;
+					total += number.doubleValue();
+				}
+				total /= params[0].length;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:721]: Also need to ensure that total itself doesn't become infinite. (could technically allow bigger sums if you divide prior to adding each number, but come on, who's getting the means of 1.7e308)

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 24]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+				double first = (sorted[half - 1]).doubleValue();
+				double second = (sorted[half]).doubleValue();
+				double median = (first+second)/2;
+				return new Number[]{median};
+			}
+		})
+			.description(
+				"Get the middle value of a sorted list of numbers. "
+				+ "If the list has an even number of values, the median is the average of the two middle numbers.",
+				"You cannot get the median of a set of numbers that includes infinity or NaN."
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:774]: wellll, you can get the median of sets that include infinity, just not NaN.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 25]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+					result *= i;
+				}
+				if (negative)
+					result *= -1d;
+				return new Number[]{result};
+			}
+		})
+			.description(
+				"Get the factorial of a number. ",
+				"Getting the factorial of any number above 20 goes into the Double range leading to floating points.",
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:813]:
```suggestion
				"Getting the factorial of any number above 21 will return an approximation, not an exact value.",
```
21! happens to be precisely representable by doubles, but 22! is not. I did not check for FP error due to the multiplication though, you should double check that in the tests (focus on 21/20/19 etc)

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 26]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+			.examples(
+				"root(2, 4) = 2 # same as sqrt(4)",
+				"root(4, 16) = 2",
+				"root(-4, 16) = 0.5 # same as 16^(-1/4)"
+			)
+			.since("INSERT VERSION");
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("permutation", new Parameter[]{
+			new Parameter<>("options", DefaultClasses.NUMBER, true, null),
+			new Parameter<>("size", DefaultClasses.NUMBER, true, null)
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:853]: I'd rather see this called `selected` or similar. The current names make it unclear which is which, though the description does explain.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 27]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+		})
+			.description("Calculates the <i>n</i>th root of a number.")
+			.examples(
+				"root(2, 4) = 2 # same as sqrt(4)",
+				"root(4, 16) = 2",
+				"root(-4, 16) = 0.5 # same as 16^(-1/4)"
+			)
+			.since("INSERT VERSION");
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("permutation", new Parameter[]{
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:851]: i'm gonna be pedantic here and say this should be `permutations`, plural, since it's counting the number of them, not actually producing a permutation.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 28]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+				for (double i = options; i > options - size; i--) {
+					if (result.isInfinite() || result.isNaN())
+						break;
+					result *= i;
+				}
+				return new Number[]{result};
+			}
+		})
+			.description(
+				"Get the number of possible ordered arrangements from 1 to 'options' with each arrangement having a size equal to 'size'",
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:876]: I'd like a line explaining the diff between perm and comb, maybe something like `(3, 2) and (2, 3) are considered the same combination, but two different permutations` or something along that line.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 29]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+				"A permutation with 3 options and an arrangement size of 2 returns 6: (1, 2), (1, 3), (2, 1), (2, 3), (3, 1), (3, 2)"
+			)
+			.examples(
+				"permutation(10, 2) = 90",
+				"permutation(10, 4) = 5040",
+				"permutation(size of {some list::*}, 2)"
+			)
+			.since("INSERT VERSION");
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("combination", new Parameter[]{
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:887]: likewise re: naming

File: src/main/java/ch/njol/util/Math2.java [hunk 30]
```diff
@@ -1,11 +1,8 @@
 package ch.njol.util;
 
 import ch.njol.skript.Skript;
-import org.jetbrains.annotations.ApiStatus.ScheduledForRemoval;
 import org.jetbrains.annotations.ApiStatus;
 
-import java.util.Arrays;
-
```
Comment (sovdeeth) [inline src/main/java/ch/njol/util/Math2.java:8]: shouldn't be touched but whatever

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 31]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+							break;
+						bottom *= i;
+					}
+					return new Number[]{top/bottom};
+				}
+			})
+			.description(
+				"Get the number of possible sets from 1 to 'options' with each set having a size equal to 'size'",
+				"For example, a combination with 3 options and a set size of 1, returns 3: (1), (2), (3)",
+				"A combination of 3 options with a set size of 2 returns 3: (1, 2), (1, 3), (2, 3)"
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:920]: These two functions should also have a disclaimer about approximations too, like factorial.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 32]
```diff
@@ -711,6 +705,227 @@ public String[] executeSimple(Object[][] params) {
...
+							return new Number[]{top};
+						top *= i;
+					}
+					Double bottom = size;
+					for (double i = size - 1; i > 1; i--) {
+						if (bottom.isInfinite() || bottom.isNaN())
+							break;
+						bottom *= i;
+					}
+					return new Number[]{top/bottom};
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:914]: Some comments would be appreciated in these functions

File: src/test/skript/tests/syntaxes/functions/factorial.sk [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+	assert factorial(1) = 1 with "factorial of 1 failed"
+	assert factorial(2) = 2 with "factorial of 2 failed"
+	assert factorial(3) = 6 with "factorial of 3 failed"
+	assert factorial(4) = 24 with "factorial of 4 failed"
+	assert factorial(5) = 120 with "factorial of 5 failed"
+	assert factorial(10) = 3628800 with "factorial of 10 failed"
+	assert factorial(100) = 93326215443944180000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 with "factorial of 100 failed"
+
+	assert factorial(500) = infinity value with "factorial of 500 failed"
+	assert factorial(-500) = (-1 * infinity value) with "factorial of -500 failed"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/functions/factorial.sk:12]:
```suggestion
	assert factorial(-500) = (-infinity value) with "factorial of -500 failed"
```

File: src/test/skript/tests/syntaxes/functions/root.sk [hunk 34]
```diff
@@ -0,0 +1,6 @@
+test "root function":
+	assert root(2, 4) = 2 with "root^2 of 4 failed"
+	assert root(4, 16) = 2 with "root^4 of 16 should be 2"
+	assert root(-4, 16) = 0.5 with "root^-4 of 16 should be 0.5"
+
+	assert root(0, 0) is not set with "root^0 should be null"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/functions/root.sk:6]:
```suggestion
	assert root(2, 4) = 2 with "root 2 of 4 failed"
	assert root(4, 16) = 2 with "root 4 of 16 should be 2"
	assert root(-4, 16) = 0.5 with "root -4 of 16 should be 0.5"

	assert root(0, 0) is not set with "root 0 should be null"
```
for clarity, i thought it was saying x^0 should be null, when it should be 1

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 35]
```diff
@@ -711,6 +705,252 @@ public String[] executeSimple(Object[][] params) {
...
+		}, DefaultClasses.NUMBER, true) {
+			@Override
+			public Number @Nullable [] executeSimple(Object[][] params) {
+				AtomicBoolean invalid = new AtomicBoolean(false);
+				// median requires the numbers to be sorted from lowest to biggest
+				Number[] sorted = Arrays.stream(params[0])
+					.filter(object -> {
+						if (!(object instanceof Number number))
+							return false;
+						if (number instanceof Double doubleNumber && doubleNumber.isNaN()) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:749]: I can't think of an exact case this happening but what if the number is a Float? Probably best to just call doubleValue and see if that's nan

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 36]
```diff
@@ -711,6 +705,252 @@ public String[] executeSimple(Object[][] params) {
...
+		}, DefaultClasses.NUMBER, true) {
+			@Override
+			public Number @Nullable [] executeSimple(Object[][] params) {
+				Double number = ((Number) params[0][0]).doubleValue();
+				boolean negative = false;
+				if (number < -170) {
+					return new Number[]{Double.NEGATIVE_INFINITY};
+				} else if (number < 0) {
+					negative = true;
+					number = Math.abs(number);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:799]: isn't factorial undefined for negative inputs?

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 25 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java]: unresolved, tests still needed

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 36 above)
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:799]: Well, according to my TI-84 calculator, negative factorials  are just treated at -1 *  #!

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:799]: That's because you're falling afoul of order of operations. ! has precedent so it calculates `-(x!)`.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 37]
```diff
@@ -711,6 +705,246 @@ public String[] executeSimple(Object[][] params) {
...
+			public Number @Nullable [] executeSimple(Object[][] params) {
+				Double total = 0d;
+				int length = params[0].length;
+				for (int i = 0; i < length; i++) {
+					Number number = (Number) params[0][i];
+					if (Double.isInfinite(number.doubleValue()) || Double.isNaN(number.doubleValue()))
+						return null;
+					if (total.isInfinite() || total.isNaN())
+						return null;
+					total += (number.doubleValue() - total) / (i + 1);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:722]: is there any reason you chose this technique instead of `+= number.doubleValue() / length`?

**Review by sovdeeth — APPROVED** [review]
seems good
